### DISCUSSION TITLE
Multiple request examples not working using pydantic

### DISCUSSION THREAD

**Discussion by Xmaster6y** [fastapi/fastapi#9794] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

from pydantic import BaseModel
from typing import Optional

class CustomRequest1(BaseModel):
    field1: Optional[str] = None
    field2: Optional[int] = None
    
    class Config:
        schema_extra = {
            "examples": [
                {
                    "summary": "Only field1",
                    "value": {
                        "field1": "value1",
                    }
                },
                {
                    "summary": "Only field2",
                    "value": {
                        "field2": 2,
                    }
                }
            ]
        }
        
class CustomRequest2(BaseModel):
    field1: Optional[str] = None
    field2: Optional[int] = None
    
    class Config:
        schema_extra = {
            "examples": [
                {
                    "field1": "value1",
                },
                {
                    "field2": 2,
                }
            ]
        }

class CustomRequest3(BaseModel):
    field1: Optional[str] = None
    field2: Optional[int] = None
        
app = FastAPI()

@app.post("/test1")
def test1(request: CustomRequest1):
    return request

@app.post("/test2")
def test2(request: CustomRequest2):
    return request

@app.post(
    "/test3",
    openapi_extra={
        "requestBody": {
            "content": {
                "application/json": {
                    "schema": {
                        "$ref": "#/components/schemas/CustomRequest3"
                        },
                    "examples": [
                        {
                            "summary": "Only field1",
                            "value": {
                                "field1": "value1",
                            }
                        },
                        {
                            "summary": "Only field2",
                            "value": {
                                "field2": 2,
                            }
                        }
                    ]
                }
            },
            "required": True,
        }
    }
          )
def test3(request: CustomRequest3):
    return request
```

### Description

The requests examples are not well documented in the first two cases. The third is working (even though a little buggy) but seems overkill.

Does this behaviour comes from Fastapi, or should I try elsewhere?

### Operating System

Linux

### FastAPI Version

0.99.1

### Python Version

3.9.16 / 3.10.11

**Comment by tiangolo** [comment, marked as answer]
Hello there!

Yep, this changed in FastAPI 0.99.0 with OpenAPI 3.1.0 and JSON Schema 2020-12.

For the first case, `examples` are now a list of pure examples without extra metadata. About the display/render of the examples, it would be in Swagger UI.

For more details, check the new docs: https://fastapi.tiangolo.com/tutorial/schema-extra-example/

In particular the technical details explaining the change: https://fastapi.tiangolo.com/tutorial/schema-extra-example/#technical-details

---

**2023-08-26 edit**: There's a new parameter `openapi_examples` just for this, to allow declaring examples that show up in Swagger UI while it implements support for multiple JSON Schema examples. It's available now in FastAPI 0.103.0, the new docs are here: https://fastapi.tiangolo.com/tutorial/schema-extra-example/#openapi-specific-examples 🎉

**Reply by Xmaster6y** [reply]
Thanks for the reply. I have already read the links provided, and I am (and was) still confused about this list of examples, only printing the first example (at least for me). Since the third example works ~well, I struggle to understand why this would be a problem with Swagger UI, but I'll dig into this further.

**Reply by tiangolo** [reply]
Yep, it's a bit confusing (and it was for me as well). There are two `examples` fields in OpenAPI, one outside JSON Schema, this is a dict, and one inside JSON Schema (part of the JSON Schema spec), this one is a list.

The one inside of JSON Schema is new, so it seems Swagger UI still doesn't support showing all the examples. But at the same time, this is the one that is most consistent with Pydantic and JSON Schema. But as you are the second person I see struggling with this, I'm considering adding an option to pass the other examples, and not just the JSON Schema ones... I don't want to add more complexity and more stuff for users to learn and understand, but maybe this is worth it 🤔

**Reply by Xmaster6y** [reply]
I see the problem; thanks for the clarification. I am unsure if this is worth it since having both will still be confusing and would only be temporary, I guess. Moreover, seeing that using openapi_extra works.

Btw I understand why you say the new `examples` field is more consistent, but it is poorly documented (not much info between [openapi spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#schemaObject) or [json schema](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00)). What's more, for it to be a list doesn't feel right unless you want to create the combination of all possible values, which would grow exponentially.

**Reply by PHvL** [reply]
I would strongly support reintroducing the non-JSON-schema examples. The possibility to describe the example values is very helpful. It allows us to use the Swagger-UI as the frontend for many small tools – it is much easier to ask the users to select from the drop-down with human-readable explanations than to type in a numerical ID.

**Reply by accountForIssues** [reply]
I would also support allowing other types of examples. Even though the new examples list looks neat, I think a lot of people have use cases where they would like to describe the different examples.

I use OpenAPI explorer instead of Swagger UI and currently the examples are broken due to this change. https://github.com/Rhosys/openapi-explorer/issues/163

**Reply by mikeedjones** [reply]
I think the [relevant section of the docs here](https://fastapi.tiangolo.com/tutorial/schema-extra-example/#examples-in-the-docs-ui) doesn't reflect the current implementation of examples in fast API and could be misleading for users. The tutorial code renders the following swaggerUI:

<img width="500" alt="image" src="https://github.com/tiangolo/fastapi/assets/4087139/007c1496-d8ee-4783-b62a-e3087dcad637">

I think this is an issue on the swagger side since they added [support for examples recently](https://github.com/swagger-api/swagger-ui/pull/8908).

**Reply by desilinguist** [reply]
I wholeheartedly second the option of support reintroducing the non-JSON-schema example as @PHvL suggested. This ability is crucial for our apps and multiple examples do not render for me current no matter what I try. I always see only the first example like @mikeedjones.

**Reply by austinorr** [reply]
@tiangolo Is it possible to revisit this? I do think OpenAPI v3.1 uses JSON Schema for the [schema](https://spec.openapis.org/oas/v3.1.0#schema-object) elements (and they _do_ have examples as lists), but not for the [Media Type](https://spec.openapis.org/oas/v3.1.0#media-type-object) which uses a map rather than a list for the examples. Swagger UI uses the examples from the Media Type to render the picker if the examples are included as a dict. 

The issue is that the `Body` kwargs are/were used for both creating the 'schema' _and_ for creating the Media Type. I opened a PR (#9866) with more discussion and a minimal start on a solution that works for endpoints that expect a `Body`.

**Reply by tiangolo** [reply]
Thanks all for the discussion and @austinorr for the PR!

---

There's a new parameter `openapi_examples` just for this, to allow declaring examples that show up in Swagger UI while it implements support for multiple JSON Schema examples. It's available now in FastAPI 0.103.0, the new docs are here: https://fastapi.tiangolo.com/tutorial/schema-extra-example/#openapi-specific-examples 🎉

**Comment by hochstibe** [comment]
Should  `Header(examples=["not working"])` work #9815  ? I think so according to the documentation (in the [docs](https://fastapi.tiangolo.com/tutorial/schema-extra-example/#pydantic-and-fastapi-examples) is more about pydantic, but it says, it should be included in the JSON schema now.

**Reply by accountForIssues** [reply]
You could directly check the spec (openapi.json) file to see if examples are being included.

In all my use cases, the spec included the examples fine but Swagger UI didn't display examples properly. Redoc worked for some examples, mainly Query/Path. And OpenAPI explorer (which I mainly use) isn't working for most examples.

So, it seems like not all tools support the new examples field.

**Reply by hochstibe** [reply]
You are right, `"examples":["not working"]` is correctly added to the openpi.json.

Then my issue is, that the `example` parameter is marked as deprecated in FastAPI, even though it is still required for a correct documentation (included in FastAPI, not an external one)

**Reply by MrCreosote** [reply]
`Path(examples=["foo"]` doesn't seem to work with OpenAPI either. `Path(example="foo")` does

**Reply by MrCreosote** [reply]
https://github.com/swagger-api/swagger-ui/issues/9045

**Comment by austinorr** [comment]
I whipped up a proposed solution to this in this PR (#9866) if anyone wants to pull it down and give it a go. I also helped get this implemented in v0.63 (#1267)

**Comment by asendia** [comment]
Hi @Xmaster6y, thanks for providing the example no. 3, you saved me a lot of time.

I found a hacky way to fix the UI select bug:
```py
from fastapi import FastAPI

from pydantic import BaseModel
from typing import Optional

class CustomRequest3(BaseModel):
    field1: Optional[str] = None
    field2: Optional[int] = None
        
app = FastAPI()

@app.post(
    "/test3",
    openapi_extra={
        "requestBody": {
            "content": {
                "application/json": {
                    "schema": {
                        "$ref": "#/components/schemas/CustomRequest3"
                        },
                    "examples": {
                        "1": {
                            "summary": "Only field1",
                            "value": {
                                "field1": "value1",
                            }
                        },
                        "2": {
                            "summary": "Only field2",
                            "value": {
                                "field2": 2,
                            }
                        }
                    }
                }
            },
            "required": True,
        }
    }
          )
def test3(request: CustomRequest3):
    return request
```

Tested on:
- Python 3.11.4
- fastapi==0.100.0
- pydantic==2.0.3

**Reply by austinorr** [reply]
I came across this too, and also found @Xmaster6y's example instructive. While working on #9866 I worked this up into a reusable function for getting examples into the Body, but it doesn't work for any of the other Params (Query, Header, Path, Cookie) since the params key in OpenAPI is a list, not a dict. With more time, this might be overcome, but a robust solution would need to intercept the openapi spec after `fastapi` has built it to loop over the params and _then_ iterate the list to set the 'examples' on the matching ('in' , 'id') combination. The current codepath of the 'openapi_extra' makes it difficult to inject the examples into the correct params element.

anyway, here's a reusable function that helps with attaching examples to Bodys via the same 'openapi_extra' kwarg on the Route:

```python
from typing import Union

from fastapi import Body, FastAPI, Path
from pydantic import BaseModel

app = FastAPI()


class Item(BaseModel):
    name: str
    description: Union[str, None] = None
    price: float
    tax: Union[float, None] = None


from typing import Any, Dict

from fastapi.openapi.models import Example


def insert_openapi_requestBody_examples(
    examples: Dict[str, Dict[str, Any]], *, media_type: str = "application/json"
):
    # don't update openapi if examples is empty dict.
    if not examples:
        return {}

    template = {
        "requestBody": {
            "content": {
                media_type: {
                    "examples": {
                        k: Example(**v).model_dump() for k, v in examples.items()
                    }
                }
            },
        }
    }

    return template


@app.put(
    "/items/{item_id}",
    openapi_extra=insert_openapi_requestBody_examples(
        examples={
            "Example Item": {
                "value": {
                    "name": "Foo",
                    "description": "A very nice Item",
                    "price": 35.4,
                    "tax": 3.2,
                }
            },
            "Example Item; coerce string to float": {
                "value": {"name": "Bar", "price": "35.4"}
            },
            "Raise validation error for 'price'": {
                "value": {
                    "name": "Baz",
                    "price": "thirty five point four",
                }
            },
        }
    ),
)
async def update_item(
    *,
    item_id: int = Path(..., examples=[0, 1, 2, 3]),
    item: Item = Body(
        examples=[
            {
                "name": "Foo",
                "description": "A very nice Item",
                "price": 35.4,
                "tax": 3.2,
            },
            {
                "name": "Bar",
                "price": "35.4",
            },
            {
                "name": "Baz",
                "price": "thirty five point four",
            },
        ],
    ),
):
    results = {"item_id": item_id, "item": item}
    return results

```
